Add Conflux namespace (#124)
ChainAgnostic/namespaces: conflux/caip10.md

---
namespace-identifier: conflux-caip10
title: Conflux Namespace - Addresses
author: iosh(@iosh)
status: Draft
type: Standard
created: 2024-08-09
requires: ["CAIP-2", "CAIP-10"]
---

# CAIP-10

_For context, see the [CAIP-10][] specification._

## Rationale

The Conflux Network EOA address is a base32-encoded address.

These are derived directly from the hexadecimal-encoded public keys corresponding to the private keys controlling them, but calculated over a network identifier, address type, and checksum value, address including a distinctive network prefix ("cfx" "cfxtest" or "net[networkId]").

The Conflux core space hex address is a concatenation of a 4-bit type indicator and the rightmost 156-bit Keccak digest of the associated public key of the private key.

There are currently three types of indicators:

(0x)1: Represents the address of an EOA account.
(0x)8: Represents the address of a contract
(0x)0: Represents the address of an internal contract,

Constructing an address:

> ```

Reviewer: ```suggestion
```json
```

Author: `js` highlight is better

encode(0x1a2f80341409639ea6a35bbcab8299066109aa55, "cfx") // cfx:aarc9abycue0hhzgyrr53m6cxedgccrmmyybjgh4xg
```

> 1. Network Prefix: cfx(mainnet network ID 1029), cfxtest(testnet network ID 1), net + network ID (ID is not 1 or 1029)

Reviewer: i'm confused by the `not `1 or 1029` part? do you use the integer chainId or the string form (`cfx`, etc)?

> 1. Network Prefix: cfx(mainnet network ID 1029), cfxtest(testnet network ID 1), net + network ID (ID is not 1 or 1029)

Reviewer: ```suggestion
1. Network Prefix: 
  * `cfx` for mainnet, i.e. network ID `1029`
  * `cfxtest` for testnet, i.e. network ID `1`
  * `net{X}` where X is the integer network ID for any other network
```

>    Network Prefix: `cfx`

Reviewer: ```suggestion
   Example address from mainnet: `cfx`
```

> 2. Payload: Concatenate version-byte: concatenate the version-byte(0x00) with hex address to get a 21-byte array.

Reviewer: ```suggestion
2. Payload: concatenate the version-byte (currently `0x00`) with the 20-byte hexadecimal address to produce a byte array 21 bytes in length.
```

   Payload: `[0x00, 0x1a, 0x2f, 0x80, 0x34, 0x14, 0x09, 0x63, 0x9e, 0xa6, 0xa3, 0x5b, 0xbc, 0xab, 0x82, 0x99, 0x06, 0x61, 0x09, 0xaa, 0x55]`

> 3. Base32 encode: encode the above result left-to-right, mapping each 5-bit sequence to the corresponding ASCII character Pad to the right with zero bits(should be 2 bit 0-padding) to complete any unfinished chunk at the end.

Reviewer: ```suggestion
3. Base32 encode: encode the above result left-to-right, mapping each 5-bit sequence to the corresponding ASCII character, then pad to the right with zero bits (2-bit `0x00`-padding) to complete any unfinished chunk at the end.
```

   5-bit base32: `[0x00, 0x00, 0x0d, 0x02, 0x1f, 0x00, 0x01, 0x14, 0x02, 0x10, 0x04, 0x16, 0x07, 0x07, 0x15, 0x06, 0x14, 0x0d, 0x0d, 0x1b, 0x19, 0x0a, 0x1c, 0x02, 0x13, 0x04, 0x03, 0x06, 0x02, 0x02, 0x0d, 0x0a, 0x0a, 0x14]`

   base32-encoded: `aarc9abycue0hhzgyrr53m6cxedgccrmmy`

> 4. Checksum Prepare checksum input: data is used as the input of checksum function. It contains:

Reviewer: ```suggestion
4. Checksum: Compute the checksum input from the preceding data as follows:
```

   4.1 The lower 5 bits of each character of the network-prefix, e.g. "cfx..." becomes 0x03, 0x06, 0x18, ...

   network-prefix cfx: `0x03, 0x06, 0x18`

   4.2 A zero for the separator (5 zero bits).

   `0x00`

   4.3 The payload by chunks of 5 bits. If necessary, the payload is padded to the right with zero bits to complete any unfinished chunk at the end.

   checksum input data:
```
   //↓ network-prefix   //↓separator //↓ 5-bit base32 with zero padding
    [0x03, 0x06, 0x18,    0x00,        0x00, 0x00, 0x0d, 0x02, 0x1f, 0x00, 0x01, 0x14, 0x02, 0x10, 0x04, 0x16, 0x07, 0x07, 0x15, 0x06, 0x14, 0x0d, 0x0d, 0x1b, 0x19, 0x0a, 0x1c, 0x02, 0x13, 0x04, 0x03, 0x06, 0x02, 0x02, 0x0d, 0x0a, 0x0a, 0x14, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
```

> 5. Calculate checksum: calculate using [Bitcoin Cash checksum algorithm][]over the data

Reviewer: ```suggestion
5. Calculate checksum: calculate using [Bitcoin Cash checksum algorithm][] over the input data
```

---
namespace-identifier: conflux-caip10
title: Conflux Namespace - Addresses
author: iosh(@iosh)
status: Draft
type: Standard
created: 2024-08-09
requires: ["CAIP-2", "CAIP-10"]
---

# CAIP-10

_For context, see the [CAIP-10][] specification._

## Rationale

The Conflux Network EOA address is a base32-encoded address.

These are derived directly from the hexadecimal-encoded public keys corresponding to the private keys controlling them, but calculated over a network identifier, address type, and checksum value, address including a distinctive network prefix ("cfx" "cfxtest" or "net[networkId]").

The Conflux core space hex address is a concatenation of a 4-bit type indicator and the rightmost 156-bit Keccak digest of the associated public key of the private key.

There are currently three types of indicators:

(0x)1: Represents the address of an EOA account.

> (0x)8: Represents the address of a contract

Reviewer 2: ```suggestion
(0x)8: Represents the address of a contract.
```

> (0x)0: Represents the address of an internal contract,

Reviewer 2: ```suggestion
(0x)0: Represents the address of an internal contract.
```

---
namespace-identifier: conflux-caip10
title: Conflux Namespace - Addresses
author: iosh(@iosh)
status: Draft
type: Standard
created: 2024-08-09
requires: ["CAIP-2", "CAIP-10"]
---

# CAIP-10

_For context, see the [CAIP-10][] specification._

## Rationale

> The Conflux Network EOA address is base32-encoded address it is derived directly from the hex-encoded addresses including a distinctive network prefix ("cfx" "cfxtest" or "net[networkId]"), an optional address type, and a checksum.

Reviewer: ```suggestion
The canonical form for each Conflux Network EOA address is a base32-encoded address.
These are derived directly from the hexadecimal-encoded public keys corresponding to the private keys controlling them, but calculated over a network identifier, address type, and checksum value, and commonly displayed with a segmented prefix containing a human-readable network identifier.
```

The hex address is a concatenation of a 4-bit type indicator and the rightmost 156-bit Keccak digest of the associated public key of the private key.

## Syntax

The syntax of Conflux addresses:

```
caip10-like address:    namespace + ":" networkId + ":" + address
namespace:              conflux
network Id:             unsigned integer ranging from 1 to 4294967295.
address:                conflux address represented as a base32 encoded string
```

## Test Cases

The list of example address generated form the same hex address, composed use the [Address Format Conversion][] or [conflux-address-js][]

```
# hex address
0x1a2f80341409639ea6a35bbcab8299066109aa55

# mainnet
conflux:1029:cfx:aarc9abycue0hhzgyrr53m6cxedgccrmmyybjgh4xg

# testnet
conflux:1:cfxtest:aarc9abycue0hhzgyrr53m6cxedgccrmmy8m50bu1p

# private net
conflux:2024:net2024:aarc9abycue0hhzgyrr53m6cxedgccrmmy06bvtn67

# private net

> conflux:202408:net202408:aarc9abycue0hhzgyrr53m6cxedgccrmmywsh35uak

Reviewer: CAIP-10 doesn't allow this redundant segment, and besides, you could think of the CAIP-10 format as the _machine-readable_ version of the more human-friendly and readable version, with `net202408` converted to `conflux:202408` to facilitate multi-L1 logs and wire formats.

Another question for your profile might whether to accept, for example,
`conflux:202408:0x1a2f80341409639ea6a35bbcab8299066109aa55` as an alias for
`conflux:202408:aarc9abycue0hhzgyrr53m6cxedgccrmmywsh35uak`, or vice versa. I have no opinion, just wondering, as sometimes systems with a per-chain address expression treat the "universal" underlying key differently.

> conflux:202408:net202408:aarc9abycue0hhzgyrr53m6cxedgccrmmywsh35uak

Author: first question：The starting part of the Conflux base32 address, such as cfx, cfxtest or net[networkid], is part of the address and cannot be removed. Therefore, it is not possible to change `conflux:1029:cfx:aarc9abycue0hhzgyrr53m6cxedgccrmmyybjgh4xg` to `conflux:1029:aarc9abycue0hhzgyrr53m6cxedgccrmmyybjgh4xg`. but i think it can be changed to `conflux:cfx:aarc9abycue0hhzgyrr53m6cxedgccrmmyybjgh4xg`. the networkid because the address already contains this

> conflux:202408:net202408:aarc9abycue0hhzgyrr53m6cxedgccrmmywsh35uak

Author: The second question:  Conflux uses a unique method to calculate EOA addresses, which means that address strings generated from the same private key will typically differ between Conflux and Ethereum. This visual similarity, combined with the computational differences, makes it easy for users to confuse Conflux addresses with Ethereum addresses, potentially leading to loss of assets. Therefore, in Conflux we do not use hex addresses.  You can find more about [base32 address](https://github.com/Conflux-Chain/CIPs/blob/master/CIPs/cip-37.md)

> conflux:202408:net202408:aarc9abycue0hhzgyrr53m6cxedgccrmmywsh35uak

Reviewer: First question: sure, you could use the native prefixes as the chainIds instead of the numerical ones, by just updating the CAIP-2 profile-- if `cfx` (the string) rather than `18` (the numeral) is used to calculate the checksum, that would make more sense! But the important part is the number of segments :D 

Second question: I gathered that from the links provided, but I also gathered that the "hex address" was used internally (or in legacy contexts) so I was just trying to come up with a more explanatory header than "## hex address" to reflect that.  Feel free to change for a better one, as long as it explains how it relates to the per-chain addresses derived from it!
